internal/provider/kubernetes/httproute.go

```diff
@@ -158,21 +158,55 @@ func (r *httpRouteReconciler) Reconcile(ctx context.Context, request reconcile.R
 		// The route's namespace exists, so add it to the resource map.
 		r.resources.Namespaces.Store(nsKey.Name, ns)
 		log.Info("added namespace to resource map")
+
+		// Get the route's backendRefs from the cache. Note that a Service is the
+		// only supported kind.
+		for i := range route.Spec.Rules {
+			for j := range route.Spec.Rules[i].BackendRefs {
+				ref := route.Spec.Rules[i].BackendRefs[j]
+				if err := validateBackendRef(&ref); err != nil {
+					return reconcile.Result{}, fmt.Errorf("invalid backendRef: %w", err)
+				}
+
+				// The backendRef is valid, so get the referenced service from the cache.
+				svcKey := types.NamespacedName{Namespace: route.Namespace, Name: string(ref.Name)}
+				svc := new(corev1.Service)
+				if err := r.client.Get(ctx, svcKey, svc); err != nil {
+					if errors.IsNotFound(err) {
+						// The ref's service doesn't exist in the cache, so remove it from
+						// the resource map if it exists.
+						if _, ok := r.resources.Services.Load(svcKey); ok {
+							r.resources.Services.Delete(svcKey)
+							log.Info("deleted service from resource map")
+						}
+					}
+					return reconcile.Result{}, fmt.Errorf("failed to get service %s/%s",
+						svcKey.Namespace, svcKey.Name)
+				}
+
+				// The backendRef Service exists, so add it to the resource map.
+				r.resources.Services.Store(svcKey, svc)
+				log.Info("added service to resource map")
+			}
+		}
 	}
 
 	if !found {
 		// Delete the httproute from the resource map.
 		r.resources.HTTPRoutes.Delete(request.NamespacedName)
 		log.Info("deleted httproute from resource map")
 
-		// Delete the namespace from the resource map if no other routes exist in the namespace.
+		// Delete the Namespace and Service from the resource maps if no other
+		// routes exist in the namespace.
 		routeList = &gwapiv1b1.HTTPRouteList{}
 		if err := r.client.List(ctx, routeList, &client.ListOptions{Namespace: request.Namespace}); err != nil {
 			return reconcile.Result{}, fmt.Errorf("error listing httproutes")
 		}
 		if len(routeList.Items) == 0 {
 			r.resources.Namespaces.Delete(request.Namespace)
 			log.Info("deleted namespace from resource map")
+			r.resources.Services.Delete(request.NamespacedName)
```

Comment:
if this backend is referenced by 2 HTTPRoutes and 1 HTTPRoute is deleted, this will cause the service to be deleted which is not what we want, same goes for the Namespace deletion in the above line

Reply:
@arkodg the services and namespace maps are only deleted when no other routes exist.

Reply:
my bad, just saw the `if len(routeList.Items) == 0 {` check